core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -132,29 +132,32 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
 
         @Nullable
         private Set<EndpointAndStep> unhandledNewEndpoints;
+        private final AtomicBoolean initialized = new AtomicBoolean();
 
         RampingUpEndpointWeightSelector(EndpointGroup endpointGroup, EventExecutor executor) {
             super(endpointGroup);
             this.executor = executor;
 
-            final AtomicBoolean initialized = new AtomicBoolean();
-            endpointGroup.addListener(newEndpoints -> {
-                if (initialized.compareAndSet(false, true)) {
-                    final List<Endpoint> dedupEndpoints =
-                            new ArrayList<>(deduplicateEndpoints(newEndpoints).values());
-                    endpointSelector = new WeightedRandomDistributionEndpointSelector(dedupEndpoints);
-                    endpointsFinishedRampingUp.addAll(dedupEndpoints);
-                } else {
-                    // Use the executor so the order of endpoints change is guaranteed.
-                    executor.execute(() -> updateEndpoints(newEndpoints));
-                }
-            }, true);
-
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
```

Comment:
How about
```suggestion
            endpointGroup.whenReady().thenAccept(this::maybeUpdateEndpoints);
```
so that we don't access the volatile field?

Reply:
`endpointGroup.whenReady()` completes with the initial value. If this callback is added after updating new endpoints which are never updated, the listener is not invoked.
`endpointGroup.endpoints()` is used to acquire the latest value instead of the initial value.

Reply:
> If this callback is added after updating new endpoints 

If so the callback is invoked right away. Isn't it?

Reply:
If the callback is added after the last update, it wouldn't be invoked.